bndtools.core/src/bndtools/central/InternalPluginTracker.java

```diff
@@ -0,0 +1,135 @@
+package bndtools.central;
+
+import java.lang.reflect.Method;
+import java.util.Formatter;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+import org.osgi.framework.Bundle;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.BundleEvent;
+import org.osgi.framework.wiring.BundleRevision;
+import org.osgi.resource.Capability;
+import org.osgi.service.metatype.annotations.AttributeDefinition;
+import org.osgi.service.metatype.annotations.ObjectClassDefinition;
+import org.osgi.util.tracker.BundleTracker;
+
+import aQute.bnd.annotation.plugin.InternalPluginDefinition;
+import aQute.bnd.annotation.plugin.InternalPluginNamespace;
+import aQute.bnd.osgi.Processor;
+
+public class InternalPluginTracker extends BundleTracker<List<InternalPluginDefinition>> {
+
+	public InternalPluginTracker(BundleContext context) {
+		super(context, Bundle.ACTIVE + Bundle.STARTING, null);
+	}
+
+	@Override
+	public List<InternalPluginDefinition> addingBundle(Bundle bundle, BundleEvent event) {
+		BundleRevision revision = bundle.adapt(BundleRevision.class);
+		List<Capability> capabilities = revision.getCapabilities(InternalPluginNamespace.NAMESPACE);
+		if (capabilities.isEmpty())
+			return null;
+
+		return capabilities.stream()
+			.map(cap -> capToDef(bundle, cap))
+			.filter(Objects::nonNull)
+			.collect(Collectors.toList());
+	}
+
+	private InternalPluginDefinition capToDef(Bundle bundle, Capability cap) {
+
+		try {
+			String name = (String) cap.getAttributes()
+				.get(InternalPluginNamespace.NAME_A);
+			String key = (String) cap.getAttributes()
+				.get(InternalPluginNamespace.IMPLEMENTATION_A);
+
+			Class<?> implementation = name == null ? null : bundle.loadClass(key);
+			key = (String) cap.getAttributes()
+				.get(InternalPluginNamespace.PARAMETERS_A);
+			Class<?> configuration = key == null ? null : bundle.loadClass(key);
+
+			Object value = cap.getAttributes()
+				.get("hide");
+			boolean hide = value != null && Processor.isTrue(value.toString());
+
+			return new InternalPluginDefinition() {
+
+				@Override
+				public String getTemplate() {
+					return InternalPluginTracker.getTemplate(name, implementation, configuration);
+				}
+
+				@Override
+				public String getName() {
+					return name;
+				}
+
+				@Override
+				public Class<?> getImplementation() {
+					return implementation;
+				}
+
+				@Override
+				public Optional<Class<?>> getParameters() {
+					return Optional.ofNullable(configuration);
+				}
+
+				@Override
+				public boolean isHidden() {
+					return hide;
+				}
+			};
+		} catch (ClassNotFoundException e) {
+			return null;
+		}
+	}
+
+	static String getTemplate(String name, Class<?> implementation, Class<?> configuration) {
+		try (Formatter sb = new Formatter()) {
+			ObjectClassDefinition od = configuration != null ? configuration.getAnnotation(ObjectClassDefinition.class)
```

Comment:
Isn't `configuration.getAnnotation(ObjectClassDefinition.class)` always `null` since ObjectClassDefinition is class retention and thus not visible at runtime?